Add release notes for 0.9.1 version (#1858)
scalameta/metals: website/blog/2020-06-27-lithium.md

Sublime Text, Atom and Eclipse. Metals is developed at the
[Scala Center](https://scala.epfl.ch/) and [VirtusLab](https://virtuslab.com)
with the help from [Lunatech](https://lunatech.com) along with contributors from
the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

- Ammonite support
- new project provider
- worksheet dependency support
- Scala 2.13.3, 0.24.0, 0.25.0-RC1 support

## Ammonite support

Thanks to a great work by [alexarchambault](https://github.com/alexarchambault)
we know support [Ammonite](https://ammonite.io) script files. Ammonite enables
to create scripts that you can run easily from the command line, without the
overhead of setting build tool while still allowing you to download and use
dependencies.

The support includes most of the LSP features such as diagnostics, completions,
references, go to definition, but does not include running or debugging the
scripts. This new feature works by starting an additional Ammonite BSP server
that compiles the scripts and provides additional information about them such as
scala version or information about the added dependencies. You can change the

> default JVM parameters for the server such as `-Xmx`, which will limit the used
> memory in Metals settings.

Reviewer: ```suggestion
default JVM parameters for the server such as `-Xmx`, which will limit the used
memory. These can be added via your metals editor extension using the newly created `metals.ammoniteJvmProperties` setting.
```

Reviewer: I'm not sure on the wording here, but I thought it might be nice to be explicit about the setting for this.

Sort of related, what are you thoughts about mentioning the `metals.ammonite-start` and `metals.ammonite-stop` commands. The only reason I mention it is that we often add commands like this without mentioning them, so users may just be totally unaware of what they do if they see them. If we mention them at least they are aware they are there and they know what they do. I don't feel strongly about this, but just an idea.

Author: Good idea! I will add that!

> Each of the scripts needs to be first imported into Ammonite BSP server, which

Reviewer: ```suggestion
Each of the scripts first need to be imported into the Ammonite BSP server, which
```

can be done manually by clicking on the pop up whenever new script is opened or

> selecting automatic import at the start, which will import any open scripts.

Reviewer: Ah maybe what I mentioned about the start and stop commands could be mentioned here?

Author: I think it's more natural to mention it after the mentioning the BSP server.

Reviewer: Sure, sounds good.

> It's important to note that this is an early approach to the support, which

Reviewer: ```suggestion
It's important to note that this is an early approach to Ammonite support, which
```

might still require some improvements both on the Metals and Ammonite side. You

> learn more about the new feature or comment the in the

Reviewer: ```suggestion
can learn more about the new feature or join the discussion in the
```

[blogpost](https://contributors.scala-lang.org/t/ammonite-support-in-metals/4306)
on the Scala contributors forum.

Most notable features of Ammonite scripts are:

> - importing other scripts, Metals provide completions to select existing files
>   based on the filesystem information

Reviewer: ```suggestion
- Import other scripts: Metals provides completions to select existing files
  based on the filesystem information
```

```scala
//main.sc
import $file.hello-world
val greeting = hello-world.greeting

//hello-world.sc
val greeting = "Hello world!"
```

> - importing dependecies

Reviewer: ```suggestion
- Import dependencies
```

```scala
import $ivy.`com.lihaoyi::scalatags:0.7.0`
import scalatags.Text.all._

val rendered = div("Moo").render
```

## Worksheet dependency support

In addition to Ammonite support, which will be useful to anyone interested in
Scala scripting, we've also improved the support for Metals Scala worksheets.

> Thanks to the efforts of [olafurpg](https://github.com/olafurpg), we now are

Reviewer: ```suggestion
Thanks to the efforts of [olafurpg](https://github.com/olafurpg) in [mdoc](https://scalameta.org/mdoc/), which powers worksheets, users are now
```

Reviewer: What do you think about mentioning mdoc here? I think it's a good idea to give credit where credit is due, but also I've found that a lot of people are unaware that mdoc is how the worksheets are actually working.

Author: Sure, good idea! I haven't thought of it.

> able to provide importing dependencies inside your Scala worksheets with
> completions and diagnostics working right from the bat.

Reviewer: ```suggestion
able to import dependencies directly inside of their Scala worksheet with
completions and diagnostics working right from the start.
```

Reviewer: Just a small change to make it more "user focused".

![worksheet-imports](https://i.imgur.com/qXgdNWM.gif)

You can add them using the following syntax:

```
$dep.`organisation::artifact:version`
```

> for example:

Reviewer: ```suggestion
For example:
```

```
$dep.`com.lihaoyi::scalatags:0.9.0`
```

> `::` is the same as `%%` in sbt, which will append the current scala binary

Reviewer: ```suggestion
`::` is the same as `%%` in sbt, which will append the current Scala binary
```

version to the artifact name.

## New project provider

Creating new Scala projects is usually a bit of work to get started, which is a

> problem for both beginners and experts alike. Normally you don't want a lot of
> time to figure out the best setup for starting with a new technology or the
> language itself.

Reviewer: ```suggestion
problem for both beginners and experts alike. Normally you don't want to spend a lot of
time to figuring out the best setup for starting with a new technology or the
language itself.
```

There are great tools like [giter8](http://www.foundweekends.org/giter8/), which
is used in the `sbt new` command, that enables people to use already existing

> Scala templates found on Github such as https://github.com/scala/scala-seed.g8.

Reviewer: ```suggestion
Scala templates found on Github such as [https://github.com/scala/scala-seed.g8](https://github.com/scala/scala-seed.g8).
```

The problem is when you don't use sbt or you don't know the exact template you
want to use.

> In this version we added an additional command called `New Scala Project` that

Reviewer: ```suggestion
To address this, we've added an additional command called `New Scala Project` that
```

Reviewer: I think since we outlined the problem in the sentence before we can change this a bit to draw a line from `existing problem` ----> `our solution`.

> enables users to reuse the same mechanism in comfort of your own UI. Firstly,

Reviewer: ```suggestion
enables users to reuse the same mechanism from the comfort of your editor. Firstly,
```

you can choose from existing templates, either curated by the Metals team, ones
existing on the giter8
[wiki page](https://github.com/foundweekends/giter8/wiki/giter8-templates) or
alternatively an entirely custom templates. Then you can choose the exact
location for the new project and the name. Afterwards, Metals will run giter8

> itself and ofer to open the new project.

Reviewer: ```suggestion
to create the project and offer to open the new project.
```

> ![new-project](https://user-images.githubusercontent.com/3807253/81842954-2d1b3680-954d-11ea-934a-96fe53cafeb9.gif)

Author: Should I make a smaller gif here? I wanted to show the entire workspace properly, but it might be too big.

Reviewer: I don't think it looks too bad, but if you are worried about it, just cutting in half and only showing the top should be fine as well.

The command is also available as a button in the welcome view of the Visual
Studio Code Metals extension, which means it will be visible in case there are
no imported projects in the current workspace.

The first curated templates were choosen by the team, however we are happy to
include more so that it's easier for people to discover new amazing libraries or
frameworks.

## Miscellaneous improvements

- fix run/debug in Maven workspaces
- always run newest sbt version if none is specified
- fix find references for Scala 3 on Windows

> - pick jvm targets before other when running

Reviewer: I'm actually a bit confused at what this means.

Author: Och yeah, that wasn't too clear :sweat_smile:

- make breakpoints work properly in Scala 3 and in case of nested classes

> - exhaustive pattern matching in case of sealed upper bound types

Reviewer: ```suggestion
- exhaustive pattern matching in the case of sealed upper bound types
```

> - compile before renaming instead showing an error

Reviewer: ```suggestion
- compile before renaming instead of showing an error
```

- fix issue with symlinked source directories
- fix missing references for synthetic companion objects
- automatically add stripMargin when using newline in a multiline string
  (behaviour can be disabled in the Metals settings)

> - automatically add -J to java properties when running or debugging

Reviewer: I don't remember if this is all the time or only if the `-J` doesn't exit. It might be worth saying _if it doesn't already exist_ if that is the case.

> - add an option to choose a build tool in workspaces with multiple ones

Reviewer: ```suggestion
- add a prompt to choose the desired build tool in workspaces with multiple build files
```

- fix go to definition in for JDK 11 library sources
- keep modifiers when implementing abstract members
- fix auto imports position when using package objects and multiple package
  definitions
- code action to make single string multiline with strip margin
- fix issues in files with Scala 2.13 specific syntax like for example
  underscore numeric literals (`val num = 100_000`)
- always run mill in interactive mode

> - improvements to hover in Scala 3 files

Reviewer: I'd capitalize the first letter of all these.

The problem is when you don't use sbt or you don't know the exact template you
want to use.

To address this, we've added an additional command called `New Scala Project`
that enables users to reuse the same mechanism from the comfort of your editor.
Firstly, you can choose from existing templates, either curated by the Metals
team, ones existing on the giter8
[wiki page](https://github.com/foundweekends/giter8/wiki/giter8-templates) or
alternatively an entirely custom templates. Then you can choose the exact
location for the new project and the name. Afterwards, Metals will run giter8 to
create the project and offer to open that new project.

![new-project](https://user-images.githubusercontent.com/3807253/81842954-2d1b3680-954d-11ea-934a-96fe53cafeb9.gif)

The command is also available as a button in the welcome view of the Visual
Studio Code Metals extension, which means it will be visible in case there are
no imported projects in the current workspace.

The first curated templates were choosen by the team, however we are happy to
include more so that it's easier for people to discover new amazing libraries or
frameworks.

> ## Multiline string improvements

Author: Added another section that might be useful for people to see. Especially the code action, which might not be obvious, but is super useful. I thinking of adding a short one about the multiple build tools maybe later.

Author: @ckipp01 or do you think the short mention is enough for that feature?

Reviewer: I think this is a really good idea having the subsection. I think I'd even suggest adding a small gif and short description of the "choose build tool one". I'm of of the mindset that anything that changes the general UX of something should be mentioned, and since that one does, it might be good to add it. Then we could also mention things like:
- What scenario is this good for? Ex. When you have a project with both a build.sbt and build.sc, but the build.sc is what you use for your build definition. Previously if both were detected, it was automatically considered an sbt workspace. Now you get to choose.
- If you choose the wrong one, or change later one, you'll need to remove the `.metals/` directory to reset the decisions.

Author: Added the next section, I think this is mostly ready. Will go over it once more once we are ready to release.

- Automatically add -J if it doesn't already exist to java properties when
  running or debugging.
- Add a prompt to choose the desired build tool in workspaces with multiple
  build files
- Fix go to definition in for JDK 11 library sources.
- Keep modifiers when implementing abstract members.
- Fix auto imports position when using package objects and multiple package
  definitions.
- Fix issues in files with Scala 2.13 specific syntax like for example
  underscore numeric literals (`val num = 100_000`).
- Always run mill in interactive mode.
- Improvements to hover in Scala 3 files.

## Contributors

Big thanks to everybody who contributed to this release or reported an issue!

```
$ git shortlog -sn --no-merges v0.9.0..v0.9.1
Tomasz Godzik
Scala Steward
Alexandre Archambault
Chris Kipp
Łukasz Trojanowski
Coline Thomas
Krzysztof Bochenek
Meriam Lachkar
bettyblue
Daniel King
Naoki Takezoe
Olafur Pall Geirsson
Win Wang
Martin Duhem
ag91
faizk
Łukasz Wawrzyk

```

## Merged PRs

## [v0.9.1](https://github.com/scalameta/metals/tree/v0.9.1) (2020-06-29)

[Full Changelog](https://github.com/scalameta/metals/compare/v0.9.0...v0.9.1)

**Merged pull requests:**

> - Update mdoc to 2.2.3 and power completions with worksheet's classpath

Author: Will update the PR list once we have a tagged version ready.

```

For example:

```
$dep.`com.lihaoyi::scalatags:0.9.0`
```

`::` is the same as `%%` in sbt, which will append the current Scala binary
version to the artifact name.

## New project provider

Creating new Scala projects is usually a bit of work to get started, which is a
problem for both beginners and experts alike. Normally you don't want to spend a
lot of time figuring out the best setup for starting with a new technology or
the language itself.

There are great tools like [giter8](http://www.foundweekends.org/giter8/), which
is used in the `sbt new` command, that enables people to use already existing
Scala templates found on Github such as
[https://github.com/scala/scala-seed.g8](https://github.com/scala/scala-seed.g8).
The problem is when you don't use sbt or you don't know the exact template you
want to use.

To address this, we've added an additional command called `New Scala Project`
that enables users to reuse the same mechanism from the comfort of your editor.
Firstly, you can choose from existing templates, either curated by the Metals
team, ones existing on the giter8
[wiki page](https://github.com/foundweekends/giter8/wiki/giter8-templates) or

> alternatively an entirely custom templates. Then you can choose the exact

Reviewer: ```suggestion
alternatively entirely custom templates. Then you can choose the exact
```
or
```suggestion
alternatively an entirely custom template. Then you can choose the exact
```

---

We're happy to announce the release of Metals v0.9.1, which focused on fixing
issues with the recently released Scala 3 support, Ammonite support and a couple
of new features.

<table>
<tbody>
  <tr>
    <td>Commits since last release</td>
    <td align="center">261</td>
  </tr>
  <tr>
    <td>Merged PRs</td>
    <td align="center">96</td>
  </tr>
    <tr>
    <td>Contributors</td>
    <td align="center">16</td>
  </tr>
  <tr>
    <td>Closed issues</td>
    <td align="center">12</td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">5</td>
  </tr>
</tbody>
</table>

For full details: https://github.com/scalameta/metals/milestone/24?closed=1

Metals is a language server for Scala that works with VS Code, Vim, Emacs,
Sublime Text, Atom and Eclipse. Metals is developed at the
[Scala Center](https://scala.epfl.ch/) and [VirtusLab](https://virtuslab.com)
with the help from [Lunatech](https://lunatech.com) along with contributors from
the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

- Ammonite support
- New project provider
- Worksheet dependency support
- Multiline string improvements

> - Scala 2.13.3, 0.24.0, 0.25.0-RC1 support

Author: Thanks, haven't noticed that one!

> - Multiple workspace build tools selection

Reviewer: Super small nitpick, but we want to make sure not to confuse people into thinking this is "multiple workspace" support in any way, so I'd recommend switching this up a bit.
```suggestion
- Multiple build tool selection within a workspace
```

Author: Fixed!

## Ammonite support

Thanks to great work by [alexarchambault](https://github.com/alexarchambault) we
now support [Ammonite](https://ammonite.io) script files. Ammonite enables you
to create scripts that can easily be ran from the command line, without the
overhead of setting up a build tool while still allowing you to download and use
dependencies.

The support includes most of the LSP features such as diagnostics, completions,
references, go to definition, but does not include running or debugging the
scripts. This new feature works by starting an additional Ammonite BSP server
that compiles the scripts and provides additional information about them such as
Scala version or information about the added dependencies. You can change the
default JVM parameters for the server such as `-Xmx`, which will limit the used
memory in Metals settings. These can be added via your metals editor extension
using the newly created `metals.ammoniteJvmProperties` setting. The Ammonite
server can also be started and stopped manually using the new commands
`ammonite-start` and `ammonite-stop`.

> Each of the scripts first need to be imported into the Ammonite BSP server,

Author: Added an example and some more info about it, what do you think?

> Each of the scripts first need to be imported into the Ammonite BSP server,

Reviewer 2: wow nice I did not know you can set scala version via comment!
Perfect example :+1:

![strip-margin](https://i.imgur.com/Z2Q1glN.gif)

This behaviour can be disabled in Metals settings using
`metals.enableStripMarginOnTypeFormatting`.

Another useful improvement coming from [colineto](https://github.com/colineto)
is a new code action that can easily convert single string into the traditional
multiline string with `|` and `stripMargin` at the end:

![new-action](https://i.imgur.com/5ohRcl9.gif)

## Multiple workspace build tools selection

Previously, Metals would only pick up a single build tool in a workspace even if
in reality there were multiple available ones. To help with that we added a
prompt to choose the desired build tool in workspaces with multiple build files.

An example situation where this is needed is when you have a project with both a
`build.sbt` and `build.sc`, but the `build.sc` is what you use for your build
definition. Previously if both were detected, it was automatically considered an
Sbt workspace, however now it's possible to choose.

![multiple-build-tool](https://i.imgur.com/pAgZK7u.gif)

If you choose the wrong one, or change the main build tool later on, you'll need
to remove the .metals/ directory to reset the decisions.

> 

Reviewer: Ah I totally forgot about that! 😄 Good memory.

> 

Author: Added! Thanks!

This great new feature was contributed by [ckipp01](https://github.com/ckipp01)!

## Miscellaneous improvements

- Fix run/debug in Maven workspaces.
- Always run newest sbt version if none is specified.
- Fix find references for Scala 3 on Windows.

> - Pick jvm targets with higher priority when running without specifying the

Author: In VS Code it can be done using launch.json. You don't have to specify the target and we will try to find one for the user. I changed it to first check for the JVM ones

> - Pick jvm targets with higher priority when running without specifying the

Author: But it can be done using different editors too, so not sure how best to really explain it here. I will change `running` to `using run/debug`

  exact build target.
- Make breakpoints work properly in Scala 3 and in case of nested classes.
- Exhaustive pattern matching in the case of sealed upper bound types.
- Compile before renaming instead of showing an error.
- Fix issue with symlinked source directories.
- Fix missing references for synthetic companion objects.
- Automatically add -J if it doesn't already exist to java properties when
  running or debugging.

> - Fix go to definition in for JDK 11 library sources.

Author: Removed
